File: 20-valid-parentheses/20-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack=[]
        map={")":"(", "]":"[", "}":"{"}
        for ch in s:
            if ch not in map:
                stack.append(ch)
            elif not stack or map[ch] != stack.pop():
                return False
        return not stack

File: 20-valid-parentheses/20-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        int n= s.length();
        if(n%2==1){return false;}
        
        Map<Character, Character> map = new HashMap<Character, Character>(){{
            put(')','(');
            put(']', '[');
            put('}','{');
        }};
        
        Deque<Character> stack =new LinkedList<Character>();
        for(int i=0; i<n;i++){
            char ch=s.charAt(i);
            if(map.containsKey(ch)){
                if(stack.isEmpty()||stack.peek()!=map.get(ch)){
                    return false;
                }
                stack.pop();
            }else {stack.push(ch);}
        }
        return stack.isEmpty();
    }
}